Rework the Configuration Warning system

RedMser opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/9416
Closes https://github.com/godotengine/godot-proposals/issues/550 again
Closes https://github.com/godotengine/godot-proposals/issues/9098
For my first attempt at implementing this (which has since been reverted), see #68420 and #88182

### What this PR does

- User facing changes
    - Configuration warnings are now a more general system called **configuration info**.
    - Configuration info is a system available on all Objects.
    - Configuration info supports different severity levels, which show up as different icons: info, warning (default), error.
    - Configuration info supports being attached to a specific property, which makes the message show in the Inspector next to that property.
    - At the top of the inspector, you can now see a list of configuration info for the selected Object (see screenshots below).
- API changes
    - A new API `Object._get_configuration_info` is available which accepts a mixed Array of Dictionaries and Strings (where `"my warning"` is equivalent to `{ "message": "my warning" }`). See docs for supported Dictionary fields.
    - C++ code that did `warnings.push_back(RTR("..."))` will now instead use `CONFIG_WARNING(RTR("..."))` macro (can be suffixed with `_P` to refer to a specific property). `INFO` and `ERROR` versions can be added once they are needed.
    - No compatibility breaking changes. New APIs for everything, old stuff is left untouched but marked deprecated. Any code still using `_get_configuration_warnings` will behave the same as before.
    - The `_get_property_warning` API was removed, since it was never exposed and is now replaced with configuration info.
    - Accessibility node warnings are merged with configuration info now (through an "accessibility" boolean field). Since this API is only part of a dev build, I did not mark it as deprecated but instead removed it completely.
    - Pulled as much logic out of Node as possible, instead moving it to its own `EditorConfigurationInfo` class, so logic can be shared between different Editor UIs. In export template builds, `_get_configuration_info` is removed to reduce binary size.

### Screenshots

Scene Tree:

![image](https://github.com/godotengine/godot/assets/5117197/fa4bd194-1d68-4b4b-9587-d449b1d8ad69)

Inspector, including summary and sub-inspectors:

![382488897-0cb5dbc1-b2c1-4f4c-ab07-4fcb7fc26c4d](https://github.com/user-attachments/assets/03036e85-0d43-44f8-bbe6-822ccd7b5929)

Export warnings:

![image](https://github.com/user-attachments/assets/1e6cc4e9-ce2a-4ac0-9402-9ce2bb11716d)

### For reviewers

First commit modifies the core and editor only, while the second commit adjusts all existing configuration warnings to use the new API instead.

I've done further changes as separate (unamended) commits because they might need separate review, or possibly even reverted.

This PR is quite large so splitting some of the changes (e.g. export warnings) to separate PRs could be worth considering.

### Notes and Known Issues

Not all of these have to be resolved to merge this PR, unless deemed necessary by reviewers.

- Test configuration info system with C# and GDExtension.
- Is the name "Configuration Info" clear enough, from a user perspective? I can not think of a better name, now that it's not only used for warnings but also errors and information text... Suggestions welcome!
- Improve the UI when clicking on the icon next to a property or in the scene tree. Should use the same RichTextLabel logic as the inspector plugin does, to show severity icons and different text color based on severity. I plan to tackle this as a follow-up PR.
- Localized property name style probably does not work properly, as it needs [additional logic from EditorInspector](https://github.com/godotengine/godot/blob/78aaa5562f4559589cfa19f173ac762e13173b46/editor/editor_inspector.cpp#L3177-L3193). Would need to update the ConfigInfo struct with classname.
- None of the built in resources currently return any configuration info. This should be done as a follow-up PR.
- All configuration info is using "warning" severity, to simplify review. A follow-up PR can check each to decide whether to use info or error for some.
- If 4.5 releases without this PR, then the original accessibility warnings API has to be kept for backwards compatibility (but deprecated). Would be annoying to deprecate an API so quickly, but oh well...

RedMser committed:
Rework the Configuration Warning system

- Introduces the new configuration info system.
- Supported on all Objects.
- Different severity levels, can be on properties.
- Show summary of config info at the top of the inspector.
- Backwards compatible API (configuration warnings API still works,
  although it is marked as deprecated).

AThousandShips commented on editor/editor_configuration_info.h:
```diff
@@ -0,0 +1,56 @@
+
+#include "core/object/ref_counted.h"
+#include "core/string/ustring.h"
+#include "core/typedefs.h"
```
```suggestion
```
Unnecessary

AThousandShips commented on doc/classes/SceneTree.xml:
```diff
@@ -278,13 +278,19 @@
+		<signal name="configuration_info_changed">
+			<param index="0" name="object" type="Object" />
+			<description>
+				Emitted when the [param object]'s [code]update_configuration_warnings[/code] method is called. Only emitted in the editor.
```
```suggestion
				Emitted when the [param object]'s [code]update_configuration_info[/code] method is called. Only emitted in the editor.
```

AThousandShips commented on doc/classes/Resource.xml:
```diff
@@ -14,6 +14,34 @@
+				Each array element must either be a [String] or a [Dictionary].
+				A dictionary element must contain a key [code]message[/code] of type [String] which is shown in the user interface, as well as following optional keys:
+				- [code]property[/code] of type [String], which also shows this warning in the inspector on the corresponding property.
+				- [code]severity[/code] can be set to [code]"error"[/code], [code]"warning"[/code] or [code]"info"[/code] to change the icon shown.
```
```suggestion
				- [code]severity[/code] can be set to [code]"error"[/code], [code]"warning"[/code], or [code]"info"[/code] to change the icon shown.
```

AThousandShips commented on doc/classes/Node.xml:
```diff
@@ -36,7 +36,35 @@
+				Each array element must either be a [String] or a [Dictionary].
+				A dictionary element must contain a key [code]message[/code] of type [String] which is shown in the user interface, as well as following optional keys:
+				- [code]property[/code] of type [String], which also shows this warning in the inspector on the corresponding property.
+				- [code]severity[/code] can be set to [code]"error"[/code], [code]"warning"[/code] or [code]"info"[/code] to change the icon shown.
```
```suggestion
				- [code]severity[/code] can be set to [code]"error"[/code], [code]"warning"[/code], or [code]"info"[/code] to change the icon shown.
```

AThousandShips commented on scene/main/node.h:
```diff
@@ -675,9 +678,15 @@ class Node : public Object {
 	_FORCE_INLINE_ Viewport *get_viewport() const { return data.viewport; }
 
+#ifndef DISABLE_DEPRECATED
 	virtual PackedStringArray get_configuration_warnings() const;
```
All the inheriting methods also need the ifndef

RedMser replied:
I assume you mean the nodes overriding/calling configuration warning stuff, like Label etc.?
I'll switch to the new methods everywhere as part of this PR, so that adding `DISABLE_DEPRECATED` shouldn't be needed anywhere else.

AThousandShips replied:
Then yes assuming you replace them (but as you can see CI fails as there's methods in `Viewport` for example) but assuming you replace them it's sufficient

RedMser commented:
> It also makes me wonder whether all `_get_configuration_warnings()` methods (or at least the method bodies) should be placed behind `#ifdef TOOLS_ENABLED` to reduce binary size.

I tried this change (ifdef the entirety of `get_configuration_info`) and it seems to reduce the binary size by 100KB. Doing so for the body only would be a lot more manual labor, not sure if it's worth the hassle.

RedMser committed:
Update all Nodes to use new configuration info API

Uses the CONFIG_ macros and specifies the property where appropriate.

RedMser commented:
This PR should be ready now.

A larger UI change has been made by introducing a configuration info summary in the inspector:

![image](https://github.com/user-attachments/assets/a6e5a5f9-0e0f-4628-b568-dc17a1e928e2)

I'd like feedback on the [proposal](https://github.com/godotengine/godot-proposals/issues/9416) whether this list should show for Resource only, or also for Node like in the current implementation.

I've also rewritten the description, to remove finished TODOs and clarify what this PR exactly does.

Mickeon commented:
Quoting [my own comment on the matter](https://github.com/godotengine/godot/issues/88379#issuecomment-1948930761) last time just for the sake for others to remember it, haven't looked at the implementation yet.

Calinou commented:
> A larger UI change has been made by introducing a configuration info summary in the inspector:

Looks good! I'd make the text besides the warning icon yellow (`warning_color`) and red text besides the error icon (`error_color`). This helps the warnings/errors stand out more.

RedMser commented:
Great idea, here's how it looks after said changes:

![image](https://github.com/user-attachments/assets/64cce5da-e7e0-4d33-ae9e-aa6b24038882)

RedMser commented on core/string/ustring.h:
```diff
@@ -623,6 +624,16 @@ _FORCE_INLINE_ String ETRN(const String &p_text, const String &p_text_plural, in
+// Internal function for configuration info macros.
+Dictionary _make_configuration_info(const String &p_message, const String &p_property_name, const String &p_severity);
+
+#define CONFIG_INFO(message) warnings.push_back(_make_configuration_info(message, "", "info"));
+#define CONFIG_INFO_P(message, property_name) warnings.push_back(_make_configuration_info(message, property_name, "info"));
+#define CONFIG_WARNING(message) warnings.push_back(_make_configuration_info(message, "", "warning"));
+#define CONFIG_WARNING_P(message, property_name) warnings.push_back(_make_configuration_info(message, property_name, "warning"));
+#define CONFIG_ERROR(message) warnings.push_back(_make_configuration_info(message, "", "error"));
+#define CONFIG_ERROR_P(message, property_name) warnings.push_back(_make_configuration_info(message, property_name, "error"));
```
Looking back at my changes, not sure why I chose `ustring` for these defines.

I'd like to move these to a better suited place, but at the same time want to avoid updating the includes of every node in the repo. Suggestions?

Also while I'm at it, I should `#ifdef TOOLS_ENABLED`, since they shouldn't be used outside of a tools build anyway.

RedMser replied:
Since I had to create a new `ConfigurationInfo` class on core (due to resources being core), it was a great place to move these macros into.

akien-mga commented on core/io/resource.cpp:
```diff
@@ -37,6 +37,11 @@
 #include "core/os/os.h"
 #include "scene/main/node.h" //only so casting works
 
+#ifdef TOOLS_ENABLED
+#include "editor/editor_node.h"
+#include "editor/editor_string_names.h"
+#endif // TOOLS_ENABLED
```
Haven't reviewed details of that PR at all, but just flagging this: `core` code shouldn't depend on `editor` code, that's breaking encapsulation rules: https://github.com/godotengine/godot/issues/53295

RedMser replied:
Thanks for letting me know!
I've fixed this particular instance by inverting the coupling, via an overridable function pointer, similar to what's done in other places in the engine.
It also created a great home for the macros, so this helped in multiple ways!

RedMser commented on scene/gui/label.cpp:
```diff
@@ -369,7 +370,10 @@ PackedStringArray Label::get_configuration_warnings() const {
-				warnings.push_back(RTR("The current font does not support rendering one or more characters used in this Label's text."));
+				Dictionary warning;
+				warning["message"] = RTR("The current font does not support rendering one or more characters used in this Label's text.");
+				warning["property"] = "text"; // REDMSER TODO: Try accessing the font using its theme override path, see if that actually shows the warning in the correct place.
```
TODO before merge: If the two commits stay split, these changes to Label should be reverted, since they were just for testing.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -3671,7 +3708,12 @@ void Node::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("update_configuration_warnings"), &Node::update_configuration_warnings);
+#endif
+#ifdef TOOLS_ENABLED
+	ClassDB::bind_method(D_METHOD("update_configuration_info"), &Node::update_configuration_info);
```
Binding the method like this will make the script break in exported project. Any exposed method in non-editor class should exist in non-tool builds (usually as empty method).

RedMser replied:
Updated this one + the virtual `_get_configuration_info` method bind.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -4084,6 +4167,52 @@ void EditorInspector::_node_removed(Node *p_node) {
+		}
+	}
+
+	for (Variant child : find_children("", "EditorInspector", true, false)) {
+		Object::cast_to<EditorInspector>(child)->_configuration_info_changed(p_object);
+	}
```
You could use `propagate_call()` instead.

RedMser replied:
Fairly sure that only works for methods that are bound, since there's no way to pass it a callable. Can a method be bound without being visible to scripting, e.g. if underscore prefixed?

KoBeWi replied:
Yes, you can prefix it with underscore.

KoBeWi commented on editor/editor_configuration_info.cpp:
```diff
@@ -0,0 +1,167 @@
+#include "scene/main/node.h"
+#include "servers/text_server.h"
+
+Array EditorConfigurationInfo::get_configuration_info_dicts(Object *p_object) {
```
It might be better to introduce a struct for configuration info. You can use it directly in C++ and have a separate method for binding that takes Dictionaries and converts them into structs.

KoBeWi commented:
IMO the inspector warnings should be folded by default. Also why are they clickable?

https://github.com/user-attachments/assets/a35283b5-f125-449a-83b7-d7c97ccd8ade

KoBeWi commented on core/io/resource.cpp:
```diff
@@ -522,6 +515,28 @@ String Resource::get_id_for_path(const String &p_path) const {
+
+	Array info;
+	if (GDVIRTUAL_CALL(_get_configuration_info, info)) {
+		ret.append_array(info);
```
`ret` is empty, so you could just use it directly.

KoBeWi commented on core/object/configuration_info.h:
```diff
@@ -0,0 +1,55 @@
+#include "core/string/ustring.h"
+#include "core/variant/dictionary.h"
+
+#define CONFIG_INFO(message) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, "", "info"));
+#define CONFIG_INFO_P(message, property_name) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, property_name, "info"));
+#define CONFIG_WARNING(message) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, "", "warning"));
+#define CONFIG_WARNING_P(message, property_name) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, property_name, "warning"));
+#define CONFIG_ERROR(message) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, "", "error"));
+#define CONFIG_ERROR_P(message, property_name) warnings.push_back(ConfigurationInfo::_make_configuration_info(message, property_name, "error"));
```
I'd remove unused macros for now. They are simple enough to add and can be added as needed.

RedMser replied:
Alright, keeping only the warning macros. And yeah not sure the variable rename is worth the effort, so I'd keep it as-is for now.

KoBeWi commented on editor/editor_configuration_info.cpp:
```diff
@@ -0,0 +1,167 @@
+		Variant item = p_mixed_config_infos[i];
+		if (item.get_type() == Variant::STRING) {
+			String str = item;
+			result.push_back(convert_string_to_dict(str));
```
This method can be inlined here. It's not used anywhere else.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -206,9 +207,27 @@ void EditorProperty::_notification(int p_what) {
+					Ref<Texture2D> config_info_icon;
+
+					String severity = EditorConfigurationInfo::get_max_severity(config_info);
+					config_info_icon = get_theme_icon(EditorConfigurationInfo::get_severity_icon(severity), SNAME("EditorIcons"));
```
```suggestion
					config_info_icon = get_editor_theme_icon(EditorConfigurationInfo::get_severity_icon(severity));
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -403,6 +422,30 @@ void EditorProperty::_notification(int p_what) {
+				Ref<Texture2D> config_info_icon;
+
+				String severity = EditorConfigurationInfo::get_max_severity(config_info);
+				config_info_icon = get_theme_icon(EditorConfigurationInfo::get_severity_icon(severity), SNAME("EditorIcons"));
```
```suggestion
				config_info_icon = get_editor_theme_icon(EditorConfigurationInfo::get_severity_icon(severity));
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -206,9 +207,27 @@ void EditorProperty::_notification(int p_what) {
+				if (show_config_info_icon) {
+					Ref<Texture2D> config_info_icon;
+
+					String severity = EditorConfigurationInfo::get_max_severity(config_info);
```
```suggestion
					const String severity = EditorConfigurationInfo::get_max_severity(config_info);
```
Same with other Strings.

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -1082,6 +1160,7 @@ void EditorProperty::_bind_methods() {
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "draw_warning"), "set_draw_warning", "is_draw_warning");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "keying"), "set_keying", "is_keying");
 	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "deletable"), "set_deletable", "is_deletable");
+	ADD_PROPERTY(PropertyInfo(Variant::ARRAY, "config_info"), "set_config_info", "get_config_info");
```
Why is this a property?
Is it even useful as readonly? (for custom EditorProperties or something)

KoBeWi commented on editor/plugins/configuration_info_editor_plugin.cpp:
```diff
@@ -0,0 +1,168 @@
+
+void ConfigurationInfoList::_notification(int p_notification) {
+	switch (p_notification) {
+		case NOTIFICATION_ENTER_TREE:
```
```suggestion
```

KoBeWi commented on scene/2d/animated_sprite_2d.cpp:
```diff
@@ -571,13 +571,17 @@ StringName AnimatedSprite2D::get_animation() const {
-		warnings.push_back(RTR("A SpriteFrames resource must be created or set in the \"Sprite Frames\" property in order for AnimatedSprite2D to display frames."));
+		CONFIG_WARNING_P(
+				RTR("A SpriteFrames resource must be created or set in order for AnimatedSprite2D to display frames."),
+				"frames");
```
```suggestion
				"sprite_frames");
```

KoBeWi commented on scene/2d/remote_transform_2d.cpp:
```diff
@@ -210,15 +210,19 @@ void RemoteTransform2D::force_update_cache() {
-		warnings.push_back(RTR("Path property must point to a valid Node2D node to work."));
+		CONFIG_WARNING_P(
+				RTR("Path must point to a valid Node2D node to work."),
+				"path");
```
```suggestion
				"remote_path");
```

RedMser commented:
> IMO the inspector warnings should be folded by default.

Sure I can change that.

> Also why are they clickable?

Yeah I have a TODO to replace them with RichTextLabel, so they wrap. Had ideas like copy text to clipboard etc, but didn't implement. Itemlist isn't good for this purpose. Will fix soon.

> Using unsupported severity results in broken icon:

Maybe it should instead fallback to "warning" severity and print an error?

KoBeWi commented on scene/3d/remote_transform_3d.cpp:
```diff
@@ -210,15 +210,19 @@ void RemoteTransform3D::force_update_cache() {
-		warnings.push_back(RTR("The \"Remote Path\" property must point to a valid Node3D or Node3D-derived node to work."));
+		CONFIG_WARNING_P(
+				RTR("Path must point to a valid Node3D or Node3D-derived node to work."),
+				"path");
```
```suggestion
				"remote_path");
```

KoBeWi commented on scene/3d/sprite_3d.cpp:
```diff
@@ -1432,13 +1432,17 @@ StringName AnimatedSprite3D::get_animation() const {
-		warnings.push_back(RTR("A SpriteFrames resource must be created or set in the \"Sprite Frames\" property in order for AnimatedSprite3D to display frames."));
+		CONFIG_WARNING_P(
+				RTR("A SpriteFrames resource must be created or set in order for AnimatedSprite3D to display frames."),
+				"frames");
```
```suggestion
				"sprite_frames");
```

KoBeWi commented on scene/gui/line_edit.cpp:
```diff
@@ -2520,13 +2520,17 @@ void LineEdit::_emit_text_change() {
 	if (secret_character.length() > 1) {
-		warnings.push_back("Secret Character property supports only one character. Extra characters will be ignored.");
+		CONFIG_WARNING_P(
+				RTR("Secret Character supports only one character. Extra characters will be ignored."),
```
This warning seems unused (you can't set longer secret).

KoBeWi commented on scene/main/canvas_item.cpp:
```diff
@@ -564,7 +564,7 @@ void CanvasItem::set_z_index(int p_z) {
 	z_index = p_z;
 	RS::get_singleton()->canvas_item_set_z_index(canvas_item, z_index);
-	update_configuration_warnings();
+	update_configuration_info();
```
This should be removed. It's a leftover.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -3338,13 +3339,49 @@ PackedStringArray Node::get_configuration_warnings() const {
+		ConfigurationInfo::configuration_info_changed_func(this);
+
+		// Emit old signal for compatibility.
 		get_tree()->emit_signal(SceneStringName(node_configuration_warning_changed), this);
```
You could wrap it in DISABLE_DEPRECATED.

KoBeWi commented on scene/main/node.h:
```diff
@@ -378,7 +379,12 @@ class Node : public Object {
+#ifndef DISABLE_DEPRECATED
 	GDVIRTUAL0RC(Vector<String>, _get_configuration_warnings)
+#endif
+#ifdef TOOLS_ENABLED
```
This might break exported projects with GDExtension.

KoBeWi commented on scene/main/timer.cpp:
```diff
@@ -180,15 +180,17 @@ void Timer::_set_process(bool p_process, bool p_force) {
 
 	if (wait_time < 0.05 - CMP_EPSILON) {
-		warnings.push_back(RTR("Very low timer wait times (< 0.05 seconds) may behave in significantly different ways depending on the rendered or physics frame rate.\nConsider using a script's process loop instead of relying on a Timer for very low wait times."));
+		CONFIG_WARNING(RTR("Very low timer wait times (< 0.05 seconds) may behave in significantly different ways depending on the rendered or physics frame rate.\nConsider using a script's process loop instead of relying on a Timer for very low wait times."));
```
That should be for `wait_time` property.

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -3406,15 +3406,17 @@ Variant Viewport::gui_get_drag_data() const {
 
 	if (size.x <= 1 || size.y <= 1) {
-		warnings.push_back(RTR("The Viewport size must be greater than or equal to 2 pixels on both dimensions to render anything."));
+		CONFIG_WARNING(RTR("The Viewport size must be greater than or equal to 2 pixels on both dimensions to render anything."));
```
For `size` property.
Although looks like another dead warning. You can't set too small size.

KoBeWi commented:
Went through the changes and overall looks fine. Though I'm not sure if there is need for supporting Strings in info arrays. It's not needed for compatibility, only for convenience.

RedMser commented:
Thanks for the review KoBeWi, I did a first pass which includes all comments marked as resolved. It's a new commit so it's easier to go over just these changes. Going to do the rest in the next few days, since it's larger changes.

> I'm not sure if there is need for supporting Strings in info arrays. It's not needed for compatibility, only for convenience.

It helps ease the transition from the old to the new system (user just has to rename the method), but indeed it wouldn't strictly be needed and currently comes with some extra complexity on the editor. I don't mind keeping it this way, or restricting the API to `Array[Dictionary]`.

RedMser commented:
OK new changes, as new commits to help make review easier.

- Replaced ItemList with RichTextLabel for the UI.
- Introduced `ConfigurationInfo` struct which is used everywhere internally, while the scripting API still uses Dictionary/String as before. This made the code a lot nicer ^^
- I changed `get_configuration_info` (the built-in one, not the one exposed to scripting) to use `ConfigurationInfo` struct, and also went and renamed `warnings` -> `infos` inside.
- Some validation was added, like printing error if message is empty, severity is unknown, property does not exist. How to make them not spam? Using `_ONCE` prints would work, but then the "property missing" error would only show once, even if you rename to another still-not-existing property...

> IMO the inspector warnings should be folded by default.

Each time you edit a property, inspector plugins are recreated, so it collapses automatically all the time. But the opposite (auto-expanding on each edit) is also not great.
Where should I store the collapse state so that it can be remembered? I should maybe push the same fix to the `ControlPositioningWarning` where I took most of the code from 👀

AThousandShips commented on scene/3d/physics/rigid_body_3d.cpp:
```diff
@@ -658,16 +658,18 @@ void RigidBody3D::_reload_physics_characteristics() {
-	PackedStringArray warnings = CollisionObject3D::get_configuration_warnings();
+#ifdef TOOLS_ENABLED
+LocalVector<ConfigurationInfo> RigidBody3D::get_configuration_info() const {
+	LocalVector<ConfigurationInfo> infos = CollisionObject3D::get_configuration_info();
```
```suggestion
	LocalVector<ConfigurationInfo> infos = PhysicsBody3D::get_configuration_info();
```
Use the direct parent (thought I had caught this one in my recent PR but must have missed it)

KoBeWi commented on core/io/resource.cpp:
```diff
@@ -522,6 +515,30 @@ String Resource::get_id_for_path(const String &p_path) const {
+
+	Array info;
+	if (GDVIRTUAL_CALL(_get_configuration_info, info)) {
+		for (int i = 0; i < info.size(); i++) {
```
You can resize the `ret` beforehand to avoid appending.
Also you can iterate Array with range iterator.

AThousandShips commented on scene/2d/physics/rigid_body_2d.cpp:
```diff
@@ -632,23 +632,25 @@ void RigidBody2D::_notification(int p_what) {
 	Transform2D t = get_transform();
 
-	PackedStringArray warnings = CollisionObject2D::get_configuration_warnings();
+	LocalVector<ConfigurationInfo> infos = CollisionObject2D::get_configuration_info();
```
```suggestion
	LocalVector<ConfigurationInfo> infos = PhysicsBody2D::get_configuration_info();
```

RedMser commented:
I've rebased to latest master and verified that everything still works. No problems when using the `template_release` either, `update_configuration_(info|warnings)` calls are a no-op as is intended. Not sure if there's anything left to do from my side now.

Once review is done, I will squash the commits to just have a "add the feature" commit and another "update all nodes to use the new APIs" commit. Is this okay or do you prefer one commit with everything squashed together?

KoBeWi commented:
> Maybe it should instead fallback to "warning" severity and print an error?

It could fallback to warning, but not sure about the error. It makes sense, but it's being spammed a lot.

https://github.com/user-attachments/assets/448a822e-9eea-4089-bc5f-6d50a45b8359

KoBeWi commented on core/object/configuration_info.cpp:
```diff
@@ -0,0 +1,88 @@
+
+ConfigurationInfo ConfigurationInfo::from_variant(const Variant &p_variant) {
+	if (p_variant.get_type() == Variant::STRING) {
+		return ConfigurationInfo(static_cast<String>(p_variant));
+	} else if (p_variant.get_type() == Variant::DICTIONARY) {
+		return ConfigurationInfo(static_cast<Dictionary>(p_variant));
```
```suggestion
		return ConfigurationInfo(String(p_variant));
	} else if (p_variant.get_type() == Variant::DICTIONARY) {
		return ConfigurationInfo(Dictionary(p_variant));
```
We don't really use `static_cast`, though not sure if it's the same performance-wise. There is also `p_variant.operator String()` syntax.

KoBeWi commented on editor/editor_configuration_info.cpp:
```diff
@@ -0,0 +1,141 @@
+	return max_severity;
+}
+
+String EditorConfigurationInfo::get_severity_icon(ConfigurationInfo::Severity p_severity) {
```
You can make this method return StringName and use SNAME in returns.

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -3384,13 +3385,48 @@ PackedStringArray Node::get_configuration_warnings() const {
+
+void Node::update_configuration_info() {
+	ERR_THREAD_GUARD
+#ifdef TOOLS_ENABLED
```
Thread guard can be included too.

KoBeWi commented on scene/main/viewport.cpp:
```diff
@@ -3316,16 +3316,6 @@ Variant Viewport::gui_get_drag_data() const {
-	PackedStringArray warnings = Node::get_configuration_warnings();
-
-	if (size.x <= 1 || size.y <= 1) {
-		warnings.push_back(RTR("The Viewport size must be greater than or equal to 2 pixels on both dimensions to render anything."));
```
What happened to this warning?

RedMser replied:
You wrote [a comment](https://github.com/godotengine/godot/pull/90049#discussion_r1769667037) saying it was a dead warning, I didn't see your edit on it until now. I can investigate why it doesn't show up on Window.

KoBeWi replied:
🤦‍♂️

RedMser replied:
Window size setter was missing `update_configuration_info();` call... but I think it makes more sense to disallow illegal values instead of using a warning for it.

KoBeWi replied:
Ah right, my EDIT refers to that. SubViewport is capped at 2, but Window is not. Weird tbh.

RedMser replied:
nvm there's more to it than just the missing update call, since Window's "size" field isn't the same as the Viewport's "size" field... same goes for `get_size` not being virtual/override... so it seems like Window needs its own `_get_configuration_info` to check for its own size property :/

KoBeWi commented on core/io/resource.cpp:
```diff
@@ -528,6 +521,33 @@ String Resource::get_id_for_path(const String &p_path) const {
+
+void Resource::update_configuration_info() {
+#ifdef TOOLS_ENABLED
+	if (!ConfigurationInfo::configuration_info_changed_func) {
+		return;
+	}
+	ConfigurationInfo::configuration_info_changed_func(this);
```
```suggestion
	if (ConfigurationInfo::configuration_info_changed_func) {
		ConfigurationInfo::configuration_info_changed_func(this);
	}
```
Early return doesn't make much sense here.

KoBeWi commented on editor/editor_configuration_info.cpp:
```diff
@@ -0,0 +1,141 @@
+String EditorConfigurationInfo::format_as_string(const ConfigurationInfo &p_config_info, bool p_wrap_lines, bool p_prefix_property_name) {
+	String text;
+
+	const StringName property_name = p_config_info.get_property_name();
```
```suggestion
	const StringName &property_name = p_config_info.get_property_name();
```

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -4088,6 +4167,51 @@ void EditorInspector::_node_removed(Node *p_node) {
+
+	// Removed entries.
+	for (uint32_t i = 0; i < property_configuration_info.size(); i++) {
+		if (found_indices.find(i) < 0) {
```
```suggestion
		if (!found_indices.has(i)) {
```

RedMser commented:
> It could fallback to warning, but not sure about the error. It makes sense, but it's being spammed a lot.

I changed that one to `ERR_PRINT_ONCE` but not sure about the ones in `ConfigurationInfo::ensure_valid`. I thought of changing the return type to an error enum value so that the errors could be returned as new configuration infos but didn't think about it much yet.

KoBeWi commented:
The problem with ERR_PRINT_ONCE is that it will appear only once per editor session 🤔
Seems like the property error is also spamming. Maybe a way to fix both is adding a deferred error print that can happen at most once per frame. Something like `queue_print_error`, with check for unique messages.

RedMser commented:
> adding a deferred error print that can happen at most once per frame. Something like `queue_print_error`, with check for unique messages.

I could try implementing this (maybe as a new PR so this one doesn't get even larger scope?). Couldn't this be implemented the same way as ERR_PRINT_ONCE, but by having a static variable that keeps track of the frame number that the error was printed on, and compare if it's different?

KoBeWi commented:
I'd make this "system" local to ConfigurationInfo. You just need a HashSet for errors and a method that prints all of them. Then once an error is to appear, add it to the HashSet instead and call deferred the method if HashSet was empty.

RedMser commented:
I implemented it like you suggested. It ""only"" prints 3 errors per update now, which I guess is better than before 😋

KoBeWi commented on core/object/configuration_info.cpp:
```diff
@@ -28,20 +28,40 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
 /**************************************************************************/
 
+#include "core/object/callable_method_pointer.h"
```
This shouldn't be needed.

KoBeWi commented on core/object/configuration_info.h:
```diff
@@ -0,0 +1,91 @@
+
+#ifdef TOOLS_ENABLED
+
+#include "core/object/object.h"
+#include "core/string/ustring.h"
+#include "core/variant/dictionary.h"
```
Might be possible to forward-declare these. Or if not, include only `variant.h`.

RedMser replied:
Could not forward declare, changed the includes to `variant.h` and also needed `hash_set.h`.

KoBeWi commented on core/object/configuration_info.h:
```diff
@@ -59,6 +59,11 @@ class ConfigurationInfo {
 	StringName property_name;
 	Severity severity;
 
+	static HashSet<String> queued_errors_to_print;
```
```suggestion
	inline static HashSet<String> queued_errors_to_print;
```
And remove from cpp.

KoBeWi commented on core/object/configuration_info.cpp:
```diff
@@ -28,20 +28,40 @@
+	if (queued_errors_to_print.is_empty()) {
+		callable_mp_static(&ConfigurationInfo::_print_errors_from_queue).call_deferred();
+	}
+
```
```suggestion
```
Same as below.

RedMser commented:
Thanks for the multiple reviews! I applied the latest changes and squashed everything into two commits. My recompile locally was instant, so my rebase maneuver did not accidentally swallow any changes either ^^

AThousandShips commented on core/io/resource.cpp:
```diff
@@ -539,6 +532,32 @@ String Resource::get_id_for_path(const String &p_path) const {
+
+	Array info;
+	if (GDVIRTUAL_CALL(_get_configuration_info, info)) {
+		ret.resize(info.size());
+
+		int idx = 0;
+		for (const Variant &variant : info) {
+			ret.set(idx++, ConfigurationInfo::from_variant(variant));
```
```suggestion
		ret.resize(info.size());
		ConfigurationInfo *ptrw = ret.ptrw();

		for (const Variant &variant : info) {
			*ptrw++ = ConfigurationInfo::from_variant(variant);
```

AThousandShips commented on editor/editor_configuration_info.cpp:
```diff
@@ -0,0 +1,140 @@
+
+#ifndef DISABLE_DEPRECATED
+		PackedStringArray warnings = node->get_configuration_warnings();
+		for (int i = 0; i < warnings.size(); i++) {
+			config_infos.push_back(ConfigurationInfo::from_variant(warnings[i]));
```
```suggestion
		for (const String &warning : warnings) {
			config_infos.push_back(ConfigurationInfo::from_variant(warning));
```

AThousandShips commented on editor/plugins/configuration_info_editor_plugin.h:
```diff
@@ -0,0 +1,88 @@
+
+#include "editor/editor_inspector.h"
+#include "editor/plugins/editor_plugin.h"
+#include "scene/gui/label.h"
+#include "scene/gui/margin_container.h"
+#include "scene/gui/panel_container.h"
+#include "scene/gui/texture_rect.h"
+
+class GridContainer;
+class RichTextLabel;
```
```suggestion
#include "scene/gui/margin_container.h"

class GridContainer;
class Label;
class PanelContainer;
class RichTextLabel;
class TextureRect;
```
And include in cpp

AThousandShips commented on scene/main/node.cpp:
```diff
@@ -3382,15 +3383,50 @@ PackedStringArray Node::get_configuration_warnings() const {
+
+	Array info;
+	if (GDVIRTUAL_CALL(_get_configuration_info, info)) {
+		ret.resize(info.size());
+
+		int idx = 0;
+		for (const Variant &variant : info) {
+			ret.set(idx++, ConfigurationInfo::from_variant(variant));
```
```suggestion
		ret.resize(info.size());
		ConfigurationInfo *ptrw = ret.ptrw();

		for (const Variant &variant : info) {
			*ptrw++ = ConfigurationInfo::from_variant(variant);
```

RedMser commented on editor/plugins/configuration_info_editor_plugin.cpp:
```diff
@@ -0,0 +1,187 @@
+
+	title_label->set_text(vformat(TTRN("%d Configuration Info", "%d Configuration Infos", config_infos.size()), config_infos.size()));
+
+	const Color &warning_color = get_theme_color(StringName("warning_color"), EditorStringName(Editor));
+	const Color &error_color = get_theme_color(StringName("error_color"), EditorStringName(Editor));
```
SNAME descrpition says it should only be used where performance is needed, and not to use when in doubt. So since it updates rarely, I didn't use it here.

But I can switch to using strings and having it convert implicitly, if that is preferred style.

RedMser commented on editor/editor_inspector.cpp:
```diff
@@ -4088,6 +4167,55 @@ void EditorInspector::_node_removed(Node *p_node) {
+	LocalVector<int> found_indices;
+	const Vector<ConfigurationInfo> config_info_list = EditorConfigurationInfo::get_configuration_info(object);
+
+	// New and changed entries.
+	for (const ConfigurationInfo &config_info : config_info_list) {
+		if (config_info.get_property_name().is_empty()) {
```
There is a problem here. This code was written with the assumption that the inspector only needs to be updated if a property has a name, since that's when the tiny ⚠️ next to its name can (dis)appear.

But now that the inspector plugin lists *all* info at the top, this logic must be updated to also take propertyless info into account.

Without this fix, the inspector banner at the top will miss some config info updates.

RedMser replied:
OK it was just a matter of removing the property_name check, easy fix.
I also renamed the variable from `property_configuration_info` to `configuration_info_cache` accordingly since it now caches all config infos on the node/resource.

Calinou commented:
@RedMser Could you look into rebasing this PR so I can review it?

RedMser commented:
@Calinou Updated to latest master! ^^

Calinou commented:
I get a build failure here:

```
Compiling scene/2d/physics/scu/scu_scene_2d_physics.gen.cpp ...
In file included from scene/2d/scu/scu_scene_2d_1.gen.cpp:4:
./scene/2d/light_occluder_2d.cpp: In member function 'void OccluderPolygon2D::set_polygon(const Vector<Vector2>&)':
./scene/2d/light_occluder_2d.cpp:93:9: error: 'update_configuration_warning' was not declared in this scope; did you mean 'update_configuration_info'?
   93 |         update_configuration_warning();
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         update_configuration_info
scons: *** [scene/2d/scu/scu_scene_2d_1.gen.linuxbsd.editor.x86_64.o] Error 1
scons: building terminated because of errors.
[Time elapsed: 00:00:37.33]
```

RedMser commented:
Seems like this got added a few hours after my rebase, haha. Did another, should work now.

Calinou reviewed (changes requested):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/c6c464cf9ae56e8b68620af65125dd980d0e8122), it works as expected. This is a great start :slightly_smiling_face: 

The API that's exposed seems good to me. Some feedback regarding the UI:

- The "N Configuration Infos" text should be changed to mention the type of messages it contains, so you can get better information without needing to expand it. For example:

> :x: 1 configuration error, 1 warning
> :warning: 2 configuration warnings, 1 info
> :x: 3 configuration errors, 1 warning, 2 infos
> :information_source: 2 configuration infos

- There should be visual feedback when hovering the dropdown that can be expanded (the "1 Configuration Info" part):

![image](https://github.com/user-attachments/assets/31218bf5-5cf3-49e2-8dcb-7f0f1391c9fb)

- On the above dropdown, the folding arrow should be moved to the left of the text (and it should point to the right when folded) to match usual Tree behavior. Here's a mockup:

![image](https://github.com/user-attachments/assets/f1685073-7a7b-4cf7-9943-239d7df9e61a)

- Line spacing should be increased within the expanded dropdown, as it's quite low currently (compared to the rest of the editor):

![image](https://github.com/user-attachments/assets/29e28337-5e1d-4d43-bd38-3bf037b1cdbd)

- The property name written in brackets should be capitalized using `EditorPropertyNameProcessor::get_singleton()->process_name()` (so that it follows editor settings). Right now, it always displays the original property name which is typically in snake_case, regardless of editor settings.

- The warning icon next to a property should be moved to the left slightly, so that it's centered:

![image](https://github.com/user-attachments/assets/bee91901-80ea-432e-8247-a7e031bc3a49)

Here's a mockup of how it should look like:

![image](https://github.com/user-attachments/assets/24b9df04-4a08-4d6f-9dc9-b7dd7a46b20c)

- When hovering the warning icon, a tooltip with the warning description should appear, like when you hover it in the Scene tree dock. Right now, it shows the property's documentation instead.

RedMser commented:
Thanks for your review and feedback! You should be able to press the "Compare" button to see the relevant code changes I made. Screenshot:

![image](https://github.com/user-attachments/assets/0cb5dbc1-b2c1-4f4c-ab07-4fcb7fc26c4d)

> The "N Configuration Infos" text should be changed 

Done, see `_get_summary_text`. I intentionally split the translations like this, since I assume some languages can't simply prefix/suffix the word "configuration" without changing conjugation.

Your mockup included emojis for the icons, but I decided not to add an icon for the dropdown title text. The list below already has many icons, so by not having an icon on the title, it gives it a bit of hierarchy. Let me know if you have a strong opinion on changing it.

> visual feedback when hovering the dropdown 
> the folding arrow should be moved to the left of the text (and it should point to the right when folded) 

Done. I've copied the UI from the "control positioning warning" inspector plugin, so it would make sense to update that one in a new PR as well.

> Line spacing should be increased 

Done.

> The property name written in brackets should be capitalized 

Done, although it will probably fail for "localized" property style, since there's a lot of additional logic in the inspector which would need to be pulled out (see code snippet). We would also need to update the struct to include class name.

https://github.com/godotengine/godot/blob/78aaa5562f4559589cfa19f173ac762e13173b46/editor/editor_inspector.cpp#L3177-L3193

I'd consider this a known issue for now which can be optimized later.

> The warning icon next to a property should [...] centered

Done by adding 50% extra width. I also noticed that I forgot to update the minimum size, so I added that and refactored duplicate code into helper methods.

> When hovering the warning icon, a tooltip [...]

It is a known limitation of this PR, since none of the icon buttons in `EditorProperty` have tooltips right now (insert keyframe, deletable, etc.). A follow-up PR could introduce a tooltip system for all of these buttons.

Calinou reviewed (approved):
The UI looks great now!

adamscott commented:
I would keep `Node::_get_configuration_warnings()` and similar methods for backward compatibility sake. Removing the functionality would break so many plugins.

I would make it so that if configuration info is given, then it takes priority on any configuration warning given.

RedMser commented:
@adamscott Current behavior is that both methods exist side-by-side and work, with `_get_configuration_warnings` marked as deprecated (to be removed in 5.0, which will likely break plugins anyway).
The results of both methods are merged together into one array, so nothing takes priority over the other.

I have not explicitly tested backwards compatibility for GDExtension or modules, but I would hope it works as intended when not compiling with `DISABLE_DEPRECATED`.
The idea is that compiling a GDExtension for 4.3 with config warnings API, and loading that into this PR, should work as expected.

AThousandShips commented:
4.4 is in feature freeze so this won't make it into 4.4, it is still waiting for review by teams this affects

RedMser commented:
FWIW, I'm not actively solving conflicts the moment they arise, since the PR quickly gets outdated again. It's inevitable due to how many files this PR touches.
Once someone is intending to look at this PR, I'll be happy to resolve conflicts prior to their review!

RedMser commented:
I have been made aware of an existing method `Object._get_property_warning` which is used for showing warning tooltips on exporter properties. It's currently undocumented API but  #105405 intends to make it public.

Since this PR shares similar goals, and still needs another review pass, I believe it would be useful to integrate following changes still:

- Move the configuration info methods into the Object class. Not only does that simplify code (no more type checking for methods that accept Object), but it also allows using this API on inspectors that aren't the main editor inspector, as it's no longer tied to Node/Resource only.
- Extend the property tooltip to include configuration info summary. While adding a tooltip to the *icon* is not easy, adding onto the *property's tooltip* seems to have already been possible, so hook onto that.
- Remove _get_property_warning and replace its usages with configuration info.

Note I haven't looked into the feature or other implications of these changes, but it seems like a worthwhile refactor to me if it all works like I plan.

RedMser committed:
[TODO AMEND] Merge Node._get_accessibility_configuration_warnings into config infos

The method was removed instead of deprecated, as it is only included in
dev builds thus far.

Separate commit for easier review, will squash before final merge

RedMser commented:
I've updated the PR:

- Rebased to latest master, fixing conflicts.
- Move accessibility warnings (#76829) to the configuration info system, by adding an `accessibility` bool flag. The EditorSettings are handled a bit weirdly, I'd like to improve that still.
- Update property tooltips to include configuration info.
    - This leeches onto the logic of `_get_property_warning`, since that already modifies the tooltip (see #74644).
![image](https://github.com/user-attachments/assets/ab510e65-e5ca-454d-bda3-cce10823e4b6)
(outdated screenshot, property name shouldn't be yellow. I'll upload a new one if I remember to...)

I tried my hand at moving configuration info to `Object` (to allow fully replacing `_get_property_warning` with it), but it seems like `GDVIRTUAL` and friends don't exist at this level... I'll have to dig into it further when I have a bit more time.
Theoretically this change should be doable as a follow-up PR (without breaking compat even?), but I don't see the point in reviewing changes which would become obsolete soon after. So I'm setting to WIP while I look into it.

lodetrick commented:
Where are you using GDVIRTUAL? The majority of them should be defined in the middle of Object.h just above the Object class declaration.

RedMser committed:
[TODO AMEND] Remove _get_property_warning in favor of configuration info

As it is an undocumented API, there should be no issue in removing it.

RedMser committed:
[TODO AMEND] Update config info without doing a full tree rebuild

RedMser commented:
I believe this is **ready for review once more**. Since my last comment, I've done the following:

- **Move configuration info system to `Object`**, so there is no similar code between `Node` and `Resource` anymore. Other inspectable objects, such as export options, can now use this system as well! 🎉 
- Remove `_get_property_warning` (it's undocumented API so nobody should depend on it). It was used by export option warnings, so I have moved this functionality to configuration info system:

![434991226-1e6cc4e9-ce2a-4ac0-9402-9ce2bb11716d](https://github.com/user-attachments/assets/18f658fb-f3da-47ab-a2bb-49280ab53370)

- I finally fixed the annoying issue of the inspector fully rebuilding its tree on config info changes. Now instead, only the property node itself is redrawn. So you no longer lose input focus in textboxes etc.

To anyone interested in this feature:

- I'd appreciate further code reviews! It'll make most sense to go commit-by-commit, due to how many files are changed.
- Help with testing is also wanted! In particular:
    - [ ] Using configuration info as well as the deprecated configuration warnings in C#.
    - [ ] Using configuration info in GDExtension.
    - [ ] Loading a GDExtension compiled on e.g. current master branch state (which also uses `get_configuration_warnings`!) into this PR.

I'll try to help as much as I can to get this merged, although I might only have limited capacity at times.

lodetrick commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -481,33 +432,14 @@ void SceneTreeEditor::_update_node(Node *p_node, TreeItem *p_item, bool p_part_o
 
-			p_item->add_button(0, get_editor_theme_icon(warning_icon), BUTTON_WARNING, false, TTR("Node configuration warning:") + all_warnings);
+			const String config_info_text = EditorConfigurationInfo::format_list_as_string(config_infos, false, true);
+			p_item->add_button(0, get_editor_theme_icon(config_info_icon), BUTTON_WARNING, false, TTR("Node configuration info:") + "\n\n" + config_info_text);
```
I agree with the old comments that two newlines don't look great with a single warning. Especially since the amount of newlines is inconsistent (two, then one). It would probably be best to pick one amount of newlines and stick with it.

RedMser replied:
Alright I'll change it to a single newline since you most commonly only have a single warning at once anyway.

Calinou commented:
> Move configuration info system to Object

I'm concerned this will have negative impacts on memory usage, considering Object is the base class of everything in Godot. Unfortunately, we can't even more it to RefCounted as Node doesn't inherit from RefCounted.

Maybe we need to find another way to avoid code duplication, e.g. using templates.

Bromeon commented on core/extension/gdextension_interface.h:
```diff
@@ -255,12 +255,20 @@ typedef struct {
 typedef GDExtensionBool (*GDExtensionClassPropertyCanRevert)(GDExtensionClassInstancePtr p_instance, GDExtensionConstStringNamePtr p_name);
 typedef GDExtensionBool (*GDExtensionClassPropertyGetRevert)(GDExtensionClassInstancePtr p_instance, GDExtensionConstStringNamePtr p_name, GDExtensionVariantPtr r_ret);
 typedef GDExtensionBool (*GDExtensionClassValidateProperty)(GDExtensionClassInstancePtr p_instance, GDExtensionPropertyInfo *p_property);
+typedef const GDExtensionConfigurationInfo *(*GDExtensionClassGetConfigurationInfo)(GDExtensionClassInstancePtr p_instance, uint32_t *r_count);
+typedef void (*GDExtensionClassFreeConfigurationInfo)(GDExtensionClassInstancePtr p_instance, const GDExtensionConfigurationInfo *p_list, uint32_t p_count);
```
Are these necessary as dedicated C functions, or could they simply be virtual functions on `Object` or `Node`?

RedMser replied:
I copied the approach of `_get_property_list` since it seems most similar to what I'm doing. I don't really know if all of the GDExtension boilerplate I carried over is *needed* for users to be able to implement their own configuration info. But it seems like Object can't (or doesn't) use GDVIRTUAL, but is instead extending the GDCLASS macro etc.

> or `Node`

Inspector can show any Object, which is done e.g. for Editor/ProjectSettings, export options, some tilemap stuff. So the idea here is that we could later add configuration info to more places than just Nodes/Resources.

RedMser commented:
> Maybe we need to find another way to avoid code duplication, e.g. using templates.

Making it a multi level method does have the upside that you no longer need to call the parent method by hand, which was fairly error-prone.
But otherwise I agree that *if* there are performance or memory implications, an alternative would be better to pursue (e.g. some kind of code generation for core classes). I just don't know if I can implement that myself...

I also don't know how to reliably profile Godot for regressions in memory or speed, but considering the GDCLASS methods are always on Object (even in exported builds), it can't be a free change.

RedMser commented:
> I also don't know how to reliably profile Godot for regressions in memory or speed

I've created a simple test project that allows instantiating and optionally freeing N objects (tried `Object` and `CanvasGroup`, the latter being deeper in the inheritance tree and having configuration info defined), measuring the time difference (and using task manager to measure memory impact).

With N = 100 000, I could not measure any difference in both categories, between the latest master and this PR, testing both editor and export templates on Windows.

<details>
<summary>Measurements</summary>

|                   | new+free Object (Time) | new Object (Size) | new+free CanvasGroup (Time) | new CanvasGroup (Size) |
| ----------------- | ---------------------- | ----------------- | --------------------------- | ---------------------- |
| Master (editor)   | 40ms                   | 50MiB             | 150ms                       | 237MiB                 |
| PR (editor)       | 40ms                   | 50MiB             | 150ms                       | 237MiB                 |
| Master (template) | 30ms                   | 50MiB             | 140ms                       | 249MiB                 |
| PR (template)     | 30ms                   | 50MiB             | 140ms                       | 248MiB                 |

Note that the values are very noisy, so they might be off by a little bit. But no outliers are visible.

</details>

I've included the project here, if anyone wants to experiment with it: [performance-objects.zip](https://github.com/user-attachments/files/19814574/performance-objects.zip)

I had an issue where the Windows export template executable from the CI/CD uses more idle CPU than my self-compiled one, but not sure what to make of it, since I can't profile the former without debug symbols.

akien-mga reviewed (commented):
Did a quick review pass on the code, looks pretty solid to me.
I still need to actually test the feature and review more, but I like where it's going.

RedMser commented on core/extension/gdextension_interface.h:
```diff
@@ -378,6 +386,8 @@ typedef struct {
 	GDExtensionClassPropertyGetRevert property_get_revert_func;
 	GDExtensionClassValidateProperty validate_property_func;
 	GDExtensionClassNotification2 notification_func;
+	GDExtensionClassGetConfigurationInfo get_configuration_info_func;
+	GDExtensionClassFreeConfigurationInfo free_configuration_info_func;
```
Hmm I was worried about that...

For the record, I don't know if there is any way to add virtual methods to Object without having to touch as many code locations as I did.
Since other classes can seemingly add virtual methods without much work, I might've done way more than what's actually needed to get something working? Or is Object just special? 🙃

RedMser commented:
> each warning should have an ID string associated to it, like GDScript warnings.

Sounds like a good idea to me!

> Defining an ID would be required when using the new API

I don't fully agree there. To me, every entry that wants to be dismissable by the user would need to define an ID (we could also add a `"dismissable": false` flag that can be set to change this default behavior). But omitting an ID should still be allowed, as we still support the String syntax in the new configinfo scripting API as a short-hand for using Dictionaries.

IMO, we could add ID as a required parameter to the CONFIG_* macros, so that the editor nodes all require to define it - but for the scripting API, it should remain optional.

---

@Calinou Do you think this should be done in the same PR (so we can avoid the noise of a future PR having to touch every node cpp file)? Or should we wait until the dismiss feature is actually getting implemented?

Mickeon commented:
> So omitting an ID (or using the old API) would simply result in an empty string as the ID, which the editor would use to determine that a warning is not ignorable.

Forgive my ignorance, but wouldn't it be more ideal, in this case, to generate an ID automatically from the warning string, instead?
I wouldn't want users to force themselves to think about IDs. It's somewhat laborious and pedantic.
If the ID of a dismissed warning doesn't exist anymore, it should then be removed to avoid accidentally polluting a scene with data.

Mickeon commented:
By the way, **in my opinion**, to be very, very safe, the newly exposed API in this PR should be marked as experimental.

RedMser commented:
> We'll need the API to be in place to allow dismissing when the PR is merged, so I would do it in this PR

Adding it later wouldn't be a breaking change either (if we make it optional as I suggest). But I'm fine adding it here as well, just gonna have to go through 170 messages haha.

> but wouldn't it be more ideal, [when omitting the ID], to generate an ID automatically from the warning string

I thought of hashing the message as well. However, warning messages may be translated, or they could have string formatting which makes their value change in unexpected ways. So the IDs wouldn't be guaranteed to be stable, and deleting old IDs from the storage might be a bad idea if the project is shared between users with different setups.

Generated IDs could be a fine fallback, but might cause problems down the line. So the more defensive solution is to simply disallow dismissing warnings unless you specify an ID.

> the newly exposed API in this PR should be marked as experimental

Sounds reasonable. One point to consider is that we'll likely want to move the API from dicts to structs in The Future™, so for that reason alone the API is expected to change.

RedMser committed:
[TODO AMEND] Add a code to each config warning

For easier review, this is a separate commit. Will be amended later.

RedMser commented on scene/2d/physics/joints/joint_2d.cpp:
```diff
@@ -138,7 +138,7 @@ void Joint2D::set_node_a(const NodePath &p_node_a) {
 	if (Engine::get_singleton()->is_editor_hint()) {
 		// When in editor, the setter may be called as a result of node rename.
-		// It happens before the node actually changes its name, which triggers false warning.
+		// It happens before the node actually changes its name, which triggers false config_info.
```
Find and replace mistake...

Zehir commented on scene/2d/physics/shape_cast_2d.cpp:
This could be linked to the `shape` property

Zehir reviewed (changes requested):
I really want this feature so I spent time to look at it. I am still new here so feel free to ignore any of my comments.

In addition to the comments below there is a "sync" issue on the errors. You can endup in a state where you have no errors on the scene tree but still having the error in the inspector;

<img width="1183" height="714" alt="Image" src="https://github.com/user-attachments/assets/0cd25464-18da-4897-9c7b-0c43f0d0beab" />

To reproduce this beavior ;
- Create a Node2D
- Set a random position
- Save the project
- You will get the warning icon in the scene tree but not in the inspector
- Select an other node
- Select back  the Node2D
- The warning will now be in sync
- Reset the position of the node
- Save the project
- The icon in the scene tree is gone but the warning in the inspector is still present
- Select an other node
- Select back the Node2D
- The warning in the inspector is gone

AThousandShips commented on core/object/configuration_info.cpp:
```diff
@@ -0,0 +1,112 @@
+	} else if (p_severity == "info") {
+		return ConfigurationInfo::Severity::INFO;
+	} else {
+		queue_error_print(vformat("Severity of Configuration Info must be one of \"error\", \"warning\" or \"info\", received \"%s\".", p_severity));
```
```suggestion
		queue_error_print(vformat("Severity of Configuration Info must be one of \"error\", \"warning\", or \"info\", received \"%s\".", p_severity));
```

AThousandShips commented on doc/classes/Object.xml:
```diff
@@ -349,11 +341,46 @@
+				Call [method update_configuration_info] when the messages need to be updated for this object.
+				The elements in the array returned from this method are displayed at the top of the Inspector dock, as well as an icon in the Scene dock for [Node]. The script that overrides it must be a [code]tool[/code] script.
+				Each array element must either be a [String] or a [Dictionary].
+				A dictionary element must contain a key [code]message[/code] of type [String] which is shown in the user interface. Following optional keys are supported:
```
```suggestion
				A dictionary element must contain a key [code]message[/code] of type [String] which is shown in the user interface. The following optional keys are supported:
```

RedMser commented on scene/3d/light_3d.cpp:
Compiler should optimize this away I believe, because of it being an empty method call.

gaheldev commented:
Would this allow to override warnings from the inherited class? 
For example inheriting from  CollisionShape3D, it would be nice to remove some or all warnings and replace it in the child class.

Here's a relevant issue: https://github.com/godotengine/godot/issues/68509
